### DISCUSSION TITLE
element-wise jacfwd for neural networks

### DISCUSSION THREAD

**Discussion by qsh-zh** [google/flax#2174] [Q&A] [answered]
Assume we neural network `model` built with flax, mapping from `x` to `y`, how can I get element-wise jacfwd. There is an example

```python
# x.shape (B, d_x)
y = model_fn(x) # y.shape (B, d_y)
jax.jacfwd(model_fn)(x) # (B, d_y, B, d_x) and most element are zeros
```

How can I get element-wise jacfwd for the model? The problem is simple if we `model_fn` works on non-batched data. But networks work in batch, how can I get element-wise jacfwd in (B, d_y, d_x) shape instead of (B, d_y, B, d_x)?

**Comment by cgarciae** [comment, marked as answer]
Hey! You need to calculate the per-sample jaccobian, you can do this by using `vmap` over `jacfwd` like this:
```python
# x.shape (B, d_x)
def per_sample_forward(x):
  return model_fn(x[None])[0] # expand input, squeeze output
y = jax.vmap(jax.jacfwd(per_sample_forward))(x) # (B, d_y, d_x)
```
